### PR TITLE
Elasticsearch: Persist custom value for size option in Terms Bucket Agg

### PR DESCRIPTION
**Pull request by Elfo404**
Follow up to #36168,  we need to to handle/persist custom select options also for `size` in `terms` bucket aggregation.

Instead of duplicating the logic i crated a custom basic hook to handle that logic in a more generic way. and applied it in bot `terms` and `date_hisatogram` settings editor.

i also moved out the `terms` settings editor in its own component to simplify a bit the codebase, feels more readable and clear.

### CODE REVIEW

File: public/app/plugins/datasource/elasticsearch/components/hooks/useCustomValue.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        } as T)
+  );
+};
+
+interface UseCustomValue<T extends SelectableValue> {
+  options: T[];
+  value?: T['value'];
+  onChange?: (value: T['value']) => void;
+}
+export const useCustomValue = <T extends SelectableValue>({
```
Comment (zoltanbedi) [inline public/app/plugins/datasource/elasticsearch/components/hooks/useCustomValue.ts:23]: I'm having a hard time seeing the value adding this hook. Which logic feels duplicated to you?

Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/components/hooks/useCustomValue.ts:23]: basically, when you create a custom value in a select, you also need to persist it somewhere, ie. by merging a set of default options and the custom created ones. then, when first mounting the component, the `options` prop should already have an option for what you set in `value`, otherwise the select will look like in an empty state.
also, when creating a new option you manually need to call the onchange because the select won't do it for you, i can see this being useful in some situations, but in the ES query editor this is not the case.

basically you'd have to implement what i did in https://github.com/grafana/grafana/pull/36168/files#diff-41c4e95630395cc02478c268278f324d4f8d61d3eeec23fbb857fa1402df4b9cR45-R63 for every select that allows custom values, and instead of testing it each time, this would basically allow me to implement the logic once and reuse it.

It is only implemented for `date_histogram` and `terms` aggregation editors now, but this should be done for all of the fields that allow custom values.

Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/components/hooks/useCustomValue.ts:23]: to add a bit more context, the other approach would have been to create a wrapper component that accepts the same props as the `Select` component in `@grafana/ui` and adds this logic, the reason i went for the hook approach is that i feel it's better suited for this case because i don't need to change the component behaviour, i just need to apply the same logic to it.

Comment (zoltanbedi) [reply public/app/plugins/datasource/elasticsearch/components/hooks/useCustomValue.ts:23]: Alright, I think what you said makes sense.

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.test.tsx [hunk 2]
```diff
@@ -1,42 +1,11 @@
-import { getDefaultTimeRange } from '@grafana/data';
```
Comment (Elfo404) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.test.tsx:1]: I'm wondering if it makes sense to have this test now. everything I was testing for here is now tested in `public/app/plugins/datasource/elasticsearch/components/hooks/useCreatableSelectPersistedBehaviour.test.tsx` and I feel this test is not adding any value, just artificially increasing the coverage

File: public/app/plugins/datasource/elasticsearch/components/hooks/useCreatableSelectPersistedBehaviour.test.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    render(<MyComp />);
+
+    const input = screen.getByLabelText('label') as HTMLInputElement;
+    expect(input).toBeInTheDocument();
+
+    // we open the menu
+    userEvent.click(input);
+
+    // default options don't have 1M but 1m
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/components/hooks/useCreatableSelectPersistedBehaviour.test.tsx:70]: This is probably a left-over from the DateHistogramSettingsEdittor test that should be deleted.

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.test.tsx (hunk 2 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.test.tsx:1]: I would be okay with removing it as well. As you've said, it is testing the same thing that is tested in `useCreatableSelectPersistedBehaviour`.

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.tsx [hunk 4]
```diff
@@ -42,51 +43,31 @@ interface Props {
...
+          })}
         />
       </InlineField>
 
       <InlineField label="Min Doc Count" {...inlineFieldProps}>
         <Input
           onBlur={(e) => dispatch(changeBucketAggregationSetting(bucketAgg, 'min_doc_count', e.target.value!))}
           defaultValue={
-            bucketAgg.settings?.min_doc_count || bucketAggregationConfig[bucketAgg.type].defaultSettings?.min_doc_count
+            bucketAgg.settings?.min_doc_count || bucketAggregationConfig.date_histogram.defaultSettings?.min_doc_count
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.tsx:70]: This is practically the same thing. So I am just curious, why did you decide to change it from `[bucketAgg.type]` to `date_histogram`.

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/index.tsx [hunk 5]
```diff
@@ -21,59 +21,13 @@ interface Props {
...
-            <Input
-              onBlur={(e) => dispatch(changeBucketAggregationSetting(bucketAgg, 'missing', e.target.value!))}
-              defaultValue={
-                bucketAgg.settings?.missing || bucketAggregationConfig[bucketAgg.type].defaultSettings?.missing
-              }
-            />
-          </InlineField>
-        </>
-      )}
+      {bucketAgg.type === 'terms' && <TermsSettingsEditor bucketAgg={bucketAgg} />}
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/index.tsx:28]: 🤩 So clean!

File: public/app/plugins/datasource/elasticsearch/test-helpers/render.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import React, { ComponentProps, ReactNode } from 'react';
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/test-helpers/render.tsx:1]: If you decide to remove `DateHistogramSettingsEditor` test, this can be removed as well, as it is not used anywhere else.

**Review by ivanahuckova — APPROVED** [review]
I have tested and reviewed this. All looks and works as expected! Superb job @Elfo404! ⭐  Left some comments/thoughts/minor comment thing, but this looks great and LGTM!

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.tsx (hunk 4 above)
Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.tsx:70]: because i felt it was going to be more readable. as you said, it can only be `date_histogram` so I don't see the point of using a computed property here.

in my head this the equivalent of doing `const a = 2;` instead of `const a = 1 + 1`: less overhead when reading the code.

but yeah, aside from this there's no practical reason :)

File: public/app/plugins/datasource/elasticsearch/components/hooks/useCreatableSelectPersistedBehaviour.test.tsx (hunk 3 above)
Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/components/hooks/useCreatableSelectPersistedBehaviour.test.tsx:70]: trueee! thanks!

File: public/app/plugins/datasource/elasticsearch/test-helpers/render.tsx (hunk 6 above)
Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/test-helpers/render.tsx:1]: Good point! I was considering the idea of using this in other tests as well (instead of using `ElasticsearchProvider` every time). do you think I should remove it now and reintroduce it only once I decide to refactor them?

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.tsx (hunk 4 above)
Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.tsx:70]: btw, open to revert the change if you think is better

Comment (ivanahuckova) [reply public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.tsx:70]: No I think it makes sense and is more readable. :) Just was curious!

File: public/app/plugins/datasource/elasticsearch/test-helpers/render.tsx (hunk 6 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/elasticsearch/test-helpers/render.tsx:1]: Hmmm, I would personally say so. 🙂 But I am okay with keeping it as well, as long as you are going to refactor it soon-ish. Just so it doesn't end up as forgotten file.

Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/test-helpers/render.tsx:1]: Damn, forgot about this and merged 🙏 
I'll make sure is not going to be forgotten
